README.md
# SalesforceBulk

[![Gem Version](https://badge.fury.io/rb/salesforcebulk.svg)](https://badge.fury.io/rb/salesforcebulk)
[![Tests](https://github.com/javierjulio/salesforce_bulk/actions/workflows/ci.yml/badge.svg)](https://github.com/javierjulio/salesforce_bulk/actions/workflows/ci.yml)

## Overview

SalesforceBulk is an easy to use Ruby gem for connecting to and using the [Salesforce Bulk API](http://www.salesforce.com/us/developer/docs/api_asynch/index.htm). This is a rewrite and separate release of Jorge Valdivia's salesforce_bulk gem (renamed `salesforcebulk`) with full unit tests and full API capability (e.g. adding multiple batches per job).

## Installation

Install SalesforceBulk from RubyGems:

```
gem install salesforcebulk
```

Or include it in your project's `Gemfile` with Bundler:

```ruby
gem 'salesforcebulk'
```

## Contribute

To contribute, fork this repo, create a topic branch, add changes and tests, then send a pull request. To setup the project and run tests in your fork, just do:

```
bundle install
bundle exec rake
```

## Configuration and Initialization

### Basic Configuration

When retrieving a password you will also be given a security token. Combine the two into a single value as the API treats this as your real password.

```ruby
require 'salesforce_bulk'

client = SalesforceBulk::Client.new(username: 'MyUsername', password: 'MyPasswordWithSecurtyToken')
client.authenticate
```

Optional keys include `login_host` (default is 'login.salesforce.com') and `version` (default is '24.0').

## Usage Examples

An important note about the data in any of the examples below: each hash in a data set must have the same set of keys. If you need to have logic to not include certain values simply specify a nil value for a key rather than not including the key-value pair.

### Basic Overall Example

```ruby
data1 = [{:Name__c => 'Test 1'}, {:Name__c => 'Test 2'}]
data2 = [{:Name__c => 'Test 3'}, {:Name__c => 'Test 4'}]

job = client.add_job(:insert, :MyObject__c)

# easily add multiple batches to a job
batch = client.add_batch(job.id, data1)
batch = client.add_batch(job.id, data2)

job = client.close_job(job.id) # or use the abort_job(id) method
```

### Adding a Job

When adding a job you can specify the following operations for the first argument:
- `:delete`
- `:insert`
- `:update`
- `:upsert`
- `:query`

When using the :upsert operation you must specify an external ID field name:

```ruby
job = client.add_job(:upsert, :MyObject__c, :external_id_field_name => :MyId__c)
```

For any operation you should be able to specify a concurrency mode. The default is `Parallel`. The only other choice is `Serial`.

```ruby
job = client.add_job(:upsert, :MyObject__c, :concurrency_mode => :Serial, :external_id_field_name => :MyId__c)
```

### Retrieving Job Information (e.g. Status)

The Job object has various properties such as status, created time, number of completed and failed batches and various other values.

```ruby
job = client.job_info(jobId) # returns a Job object

puts "Job #{job.id} is closed." if job.closed? # other: open?, aborted?
```

### Retrieving Info for a single Batch

The Batch object has various properties such as status, created time, number of processed and failed records and various other values.

```ruby
batch = client.batch_info(jobId, batchId) # returns a Batch object

puts "Batch #{batch.id} is in progress." if batch.in_progress?
```

### Retrieving Info for all Batches

```ruby
batches = client.batch_info_list(jobId) # returns an Array of Batch objects

batches.each do |batch|
    puts "Batch #{batch.id} completed." if batch.completed? # other: failed?, in_progress?, queued?
end
```

### Retrieving Batch Results (for Delete, Insert, Update and Upsert)

To verify that a batch completed successfully or failed call the `batch_info` or `batch_info_list` methods first, otherwise if you call `batch_result` without verifying and the batch failed the method will raise an error.

The object returned from the following example only applies to the operations: `delete`, `insert`, `update` and `upsert`. Query results are handled differently.

```ruby
results = client.batch_result(jobId, batchId) # returns an Array of BatchResult objects

results.each do |result|
    puts "Item #{result.id} had an error of: #{result.error}" if result.error?
end
```

### Retrieving Query based Batch Results

To verify that a batch completed successfully or failed call the `batch_info` or `batch_info_list` methods first, otherwise if you call `batch_result` without verifying and the batch failed the method will raise an error.

Query results are handled differently as its possible that a single batch could return multiple results if objects returned are large enough. Note: I haven't been able to replicate this behavior but in a fork by @WWJacob has [discovered that multiple results can be returned](https://github.com/WWJacob/salesforce_bulk/commit/8f9e68c390230e885823e45cd2616ac3159697ef).

```ruby
# returns a QueryResultCollection object (an Array)
results = client.batch_result(jobId, batchId)

while results.any?

    # Assuming query was: SELECT Id, Name, CustomField__c FROM Account
    results.each do |result|
    puts result[:Id], result[:Name], result[:CustomField__c]
    end

    puts "Another set is available." if results.next?

    results.next

end
```

Note: By reviewing the API docs and response format my understanding was that the API would return multiple results sets for a single batch if the query was to large but this does not seem to be the case in my live testing. It seems to be capped at 10000 records (as it when inserting data) but I haven't been able to verify through the documentation. If you know anything about that your input is appreciated. In the meantime the gem was built to support multiple result sets for a query batch but seems that will change which will simplify that method.

## Releasing

To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contribution Suggestions/Ideas

- Support for other Ruby platforms
- Clean up/reorganize tests better
- Rdocs

### Releasing

To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/javierjulio/salesforce_bulk. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](http://contributor-covenant.org) code of conduct.

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).


CHANGELOG.md
## Changelog

**4.0.4** (February 23, 2023)

* Add `exception_code` attribute to `SalesforceError` with parsed exception code ([#27](https://github.com/javierjulio/salesforce_bulk/pull/27) - thanks [@SValkanov](https://github.com/SValkanov))

**4.0.3** (February 11, 2023)

* Make `session_id` on `Client` writable ([#26](https://github.com/javierjulio/salesforce_bulk/pull/26) - thanks [@SValkanov](https://github.com/SValkanov))

**4.0.2** (September 21, 2022)

* Avoid NoMethodError exceptions on nil when authentication response is a redirect ([#24](https://github.com/javierjulio/salesforce_bulk/pull/24) - thanks [@anero](https://github.com/anero))

**4.0.1** (April 13, 2021)

* Escape special characters in password during authentication ([#23](https://github.com/javierjulio/salesforce_bulk/pull/23) - thanks [@lavaturtle](https://github.com/lavaturtle))

**4.0.0** (March 5, 2021)

* Removed YAML configuration support
* Removed ActiveSupport dependency
* Removed String monkey patch: `to_b` method (to boolean)
* Replaced Travis with GitHub Actions CI
* Added Ruby CI matrix from 2.3 to 3.0

**3.0.0** (July 19, 2017)

* Dropped old Ruby support, now requires Ruby 2.3 and up
* Dropped Travis build for Rails 4.0
* Added support for re-authenticating ([#20](https://github.com/javierjulio/salesforce_bulk/pull/20))
* Use the right tag name for total batches in job info ([#19](https://github.com/javierjulio/salesforce_bulk/pull/19))

**2.0.2** (February 20, 2017)

* Rails 5 support by relaxing ActiveSupport dependency ([#18](https://github.com/javierjulio/salesforce_bulk/pull/18))

**2.0.1** (January 24, 2017)

* Bug fix for response handling ([#17](https://github.com/javierjulio/salesforce_bulk/pull/17))

**2.0.0** (April 25, 2015)

* Dropped support for Ruby 1.8 and Ruby 1.9
* Added support for Ruby 2.0, 2.1 and 2.2
* Added support for Rails 4.0, 4.1 an 4.2
* Changed test_helper to avoid requiring test_unit (removed in Ruby 2.2)
* Replaced Test::Unit::TestCase with ActiveSupport::TestCase
* Bumped shoulda and losen dependencies on minitest
* All changes in PR's #13, #14, #15, #16 - thanks [@pschambacher](https://github.com/pschambacher)

**1.4.0** (June 1, 2014)

* Added state_message to Batch class (#11 - thanks [@bethesque](https://github.com/bethesque))

**1.3.0** (April 28, 2014)

* Added support for multiple subdomains (#10 - thanks [@lucianapazos](https://github.com/lucianapazos))
* Added dependency version requirements to gemspec

**1.2.0** (October 10, 2012)

* Added Ruby 1.8.7 support (thanks [@dlee](https://github.com/dlee))

**1.1.0** (August 20, 2012)

* Added travis setup. Support for Ruby 1.9.2 and 1.9.3 specified.
* Removed `token` property on Client object. Specify token in `password` field.
* Accepted pull request for 1.9.3 improvements.
* Description updates in README.

**1.0.0** (August 17, 2012)

* Initial public release.


test/test_helper.rb
$LOAD_PATH.unshift File.expand_path("../lib", __dir__)
require 'salesforce_bulk'
require 'minitest/autorun'
require 'minitest/reporters'
require 'mocha/minitest'
require 'webmock/minitest'

Minitest::Reporters.use!([Minitest::Reporters::SpecReporter.new])

class Minitest::Test
  def self.test(name, &block)
    test_name = "test_#{name.gsub(/\s+/, '_')}".to_sym
    defined = method_defined? test_name
    raise "#{test_name} is already defined in #{self}" if defined
    if block_given?
      define_method(test_name, &block)
    else
      define_method(test_name) do
        flunk "No implementation provided for #{name}"
      end
    end
  end
end

def api_url(client)
  "https://#{client.login_host}/services/async/#{client.version}/"
end

def bypass_authentication(client)
  client.instance_variable_set('@session_id', '123456789')
  client.instance_variable_set('@login_host', 'na9.salesforce.com')
  client.instance_variable_set('@instance_host', 'na9.salesforce.com')
end

def fixture_path(file)
  File.expand_path("../fixtures/#{file}", __FILE__)
end

def fixture(file)
  File.new(fixture_path(file)).read
end


test/lib/test_job.rb
require 'test_helper'

class TestJob < Minitest::Test

  def setup
    options = {
      :username => 'myusername',
      :password => 'mypassword'
    }

    @client = SalesforceBulk::Client.new(options)
    @job = SalesforceBulk::Job.new
    @headers = {'Content-Type' => 'application/xml', 'X-Sfdc-Session' => '123456789'}

    bypass_authentication(@client)
  end

  test "initialization from XML" do
    xml = fixture("job_info_response.xml")
    job = SalesforceBulk::Job.new_from_xml(XmlSimple.xml_in(xml, 'ForceArray' => false))

    assert_equal job.id, '750E00000004N1mIAE'
    assert_equal job.operation, 'upsert'
    assert_equal job.sobject, 'VideoEvent__c'
    assert_equal job.created_by, '005E00000017spfIAA'
    assert_equal job.created_at, DateTime.parse('2012-05-30T04:08:30.000Z')
    assert_equal job.completed_at, DateTime.parse('2012-05-30T04:08:30.000Z')
    assert_equal job.state, 'Open'
    assert_equal job.external_id_field_name, 'Id__c'
    assert_equal job.concurrency_mode, 'Parallel'
    assert_equal job.content_type, 'CSV'
    assert_equal job.queued_batches, 0
    assert_equal job.in_progress_batches, 0
    assert_equal job.completed_batches, 0
    assert_equal job.failed_batches, 0
    assert_equal job.total_batches, 0
    assert_equal job.processed_records, 0
    assert_equal job.failed_records, 0
    assert_equal job.retries, 0
    assert_equal job.api_active_processing_time, 0
    assert_equal job.apex_processing_time, 0
    assert_equal job.total_processing_time, 0
    assert_equal job.api_version, 24.0
  end

  test "state?" do
    @job.state = 'Closed'
    assert @job.state?('closed')

    @job.state = 'Closed'
    assert @job.state?('CLOSED')

    @job.state = nil
    assert !@job.state?('closed')
  end

  test "aborted?" do
    @job.state = 'Aborted'
    assert @job.aborted?

    @job.state = nil
    assert !@job.aborted?
  end

  test "closed?" do
    @job.state = 'Closed'
    assert @job.closed?

    @job.state = nil
    assert !@job.closed?
  end

  test "open?" do
    @job.state = 'Open'
    assert @job.open?

    @job.state = nil
    assert !@job.open?
  end

  test "add_job returns successful response" do
    request = fixture("job_create_request.xml")
    response = fixture("job_create_response.xml")

    stub_request(:post, "#{api_url(@client)}job").with(:body => request, :headers => @headers).to_return(:body => response, :status => 200)

    job = @client.add_job(:upsert, :VideoEvent__c, :external_id_field_name => :Id__c)

    assert_requested :post, "#{api_url(@client)}job", :body => request, :headers => @headers, :times => 1

    assert_equal job.id, '750E00000004MzbIAE'
    assert_equal job.operation, 'upsert'
    assert_equal job.sobject, 'VideoEvent__c'
    assert_equal job.created_by, '005E00000017spfIAA'
    assert_equal job.created_at, DateTime.parse('2012-05-29T21:50:47.000Z')
    assert_equal job.completed_at, DateTime.parse('2012-05-29T21:50:47.000Z')
    assert_equal job.state, 'Open'
    assert_equal job.external_id_field_name, 'Id__c'
    assert_equal job.concurrency_mode, 'Parallel'
    assert_equal job.content_type, 'CSV'
    assert_equal job.queued_batches, 0
    assert_equal job.in_progress_batches, 0
    assert_equal job.completed_batches, 0
    assert_equal job.failed_batches, 0
    assert_equal job.total_batches, 0
    assert_equal job.processed_records, 0
    assert_equal job.failed_records, 0
    assert_equal job.retries, 0
    assert_equal job.api_active_processing_time, 0
    assert_equal job.apex_processing_time, 0
    assert_equal job.total_processing_time, 0
    assert_equal job.api_version, 24.0
  end

  test "add_job raises ArgumentError if provided with invalid operation" do
    assert_raises ArgumentError do
      @client.add_job(:SomeOtherOperation, nil)
    end
  end

  test "add_job raises ArgumentError if provided with invalid key for options" do
    assert_raises ArgumentError do
      @client.add_job(:upsert, :VideoEvent__c, :non_existing_key => '')
    end
  end

  test "add_job raises ArgumentError if provided with invalid concurrency mode" do
    assert_raises ArgumentError do
      @client.add_job(:upsert, :VideoEvent__c, :concurrency_mode => 'SomeMode')
    end
  end

  test "should close job and return successful response" do
    request = fixture("job_close_request.xml")
    response = fixture("job_close_response.xml")
    job_id = "750E00000004MzbIAE"

    stub_request(:post, "#{api_url(@client)}job/#{job_id}").with(:body => request, :headers => @headers).to_return(:body => response, :status => 200)

    job = @client.close_job(job_id)

    assert_requested :post, "#{api_url(@client)}job/#{job_id}", :body => request, :headers => @headers, :times => 1

    assert_equal job.id, job_id
    assert_equal job.operation, 'upsert'
    assert_equal job.sobject, 'VideoEvent__c'
    assert_equal job.created_by, '005E00000017spfIAA'
    assert_equal job.created_at, DateTime.parse('2012-05-29T23:51:53.000Z')
    assert_equal job.completed_at, DateTime.parse('2012-05-29T23:51:53.000Z')
    assert_equal job.state, 'Closed'
    assert_equal job.external_id_field_name, 'Id__c'
    assert_equal job.concurrency_mode, 'Parallel'
    assert_equal job.content_type, 'CSV'
    assert_equal job.queued_batches, 0
    assert_equal job.in_progress_batches, 0
    assert_equal job.completed_batches, 0
    assert_equal job.failed_batches, 0
    assert_equal job.total_batches, 0
    assert_equal job.processed_records, 0
    assert_equal job.failed_records, 0
    assert_equal job.retries, 0
    assert_equal job.api_active_processing_time, 0
    assert_equal job.apex_processing_time, 0
    assert_equal job.total_processing_time, 0
    assert_equal job.api_version, 24.0
  end

  test "should abort job and return successful response" do
    request = fixture("job_abort_request.xml")
    response = fixture("job_abort_response.xml")
    job_id = "750E00000004N1NIAU"

    stub_request(:post, "#{api_url(@client)}job/#{job_id}").with(:body => request, :headers => @headers).to_return(:body => response, :status => 200)

    job = @client.abort_job(job_id)

    assert_requested :post, "#{api_url(@client)}job/#{job_id}", :body => request, :headers => @headers, :times => 1

    assert_equal job.id, job_id
    assert_equal job.operation, 'upsert'
    assert_equal job.sobject, 'VideoEvent__c'
    assert_equal job.created_by, '005E00000017spfIAA'
    assert_equal job.created_at, DateTime.parse('2012-05-30T00:16:04.000Z')
    assert_equal job.completed_at, DateTime.parse('2012-05-30T00:16:04.000Z')
    assert_equal job.state, 'Aborted'
    assert_equal job.external_id_field_name, 'Id__c'
    assert_equal job.concurrency_mode, 'Parallel'
    assert_equal job.content_type, 'CSV'
    assert_equal job.queued_batches, 0
    assert_equal job.in_progress_batches, 0
    assert_equal job.completed_batches, 0
    assert_equal job.failed_batches, 0
    assert_equal job.total_batches, 0
    assert_equal job.processed_records, 0
    assert_equal job.failed_records, 0
    assert_equal job.retries, 0
    assert_equal job.api_active_processing_time, 0
    assert_equal job.apex_processing_time, 0
    assert_equal job.total_processing_time, 0
    assert_equal job.api_version, 24.0
  end

  test "should return job info" do
    response = fixture("job_info_response.xml")
    job_id = "750E00000004N1mIAE"

    stub_request(:get, "#{api_url(@client)}job/#{job_id}").with(:body => '', :headers => @headers).to_return(:body => response, :status => 200)

    job = @client.job_info(job_id)

    assert_requested :get, "#{api_url(@client)}job/#{job_id}", :body => '', :headers => @headers, :times => 1

    assert_equal job.id, job_id
    assert_equal job.operation, 'upsert'
    assert_equal job.sobject, 'VideoEvent__c'
    assert_equal job.created_by, '005E00000017spfIAA'
    assert_equal job.created_at, DateTime.parse('2012-05-30T04:08:30.000Z')
    assert_equal job.completed_at, DateTime.parse('2012-05-30T04:08:30.000Z')
    assert_equal job.state, 'Open'
    assert_equal job.external_id_field_name, 'Id__c'
    assert_equal job.concurrency_mode, 'Parallel'
    assert_equal job.content_type, 'CSV'
    assert_equal job.queued_batches, 0
    assert_equal job.in_progress_batches, 0
    assert_equal job.completed_batches, 0
    assert_equal job.failed_batches, 0
    assert_equal job.total_batches, 0
    assert_equal job.processed_records, 0
    assert_equal job.failed_records, 0
    assert_equal job.retries, 0
    assert_equal job.api_active_processing_time, 0
    assert_equal job.apex_processing_time, 0
    assert_equal job.total_processing_time, 0
    assert_equal job.api_version, 24.0
  end

  test "should raise SalesforceError on invalid job" do
    response = fixture("invalid_job_error.xml")

    stub_request(:post, "#{api_url(@client)}job").to_return(:body => response, :status => 500)

    error = assert_raises SalesforceBulk::SalesforceError do
      @client.add_job(:upsert, :SomeNonExistingObject__c, :external_id_field_name => :Id__c)
    end

    assert_equal '500', error.error_code
    assert_equal 'Unable to find object: Video__c', error.message
    assert_equal 'InvalidJob', error.exception_code
  end

  test "should raise SalesforceError on invalid session response" do
    response = fixture("invalid_session_error.xml")

    stub_request(:post, "#{api_url(@client)}job").to_return(:body => response, :status => 400)

    error = assert_raises SalesforceBulk::SalesforceError do
      @client.add_job(:upsert, :SomeNonExistingObject__c, :external_id_field_name => :Id__c)
    end

    assert_equal '400', error.error_code
    assert_equal 'Invalid session id', error.message
    assert_equal 'InvalidSessionId', error.exception_code
  end
end


lib/salesforce_bulk/job.rb
module SalesforceBulk
  class Job

    attr_accessor :concurrency_mode
    attr_accessor :external_id_field_name
    attr_accessor :id
    attr_accessor :operation
    attr_accessor :sobject
    attr_accessor :state
    attr_accessor :created_by
    attr_accessor :created_at
    attr_accessor :completed_at
    attr_accessor :content_type
    attr_accessor :queued_batches
    attr_accessor :in_progress_batches
    attr_accessor :completed_batches
    attr_accessor :failed_batches
    attr_accessor :total_batches
    attr_accessor :retries
    attr_accessor :failed_records
    attr_accessor :processed_records
    attr_accessor :apex_processing_time
    attr_accessor :api_active_processing_time
    attr_accessor :total_processing_time
    attr_accessor :api_version

    def self.new_from_xml(data)
      job = self.new
      job.id = data['id']
      job.operation = data['operation']
      job.sobject = data['object']
      job.created_by = data['createdById']
      job.state = data['state']
      job.created_at = DateTime.parse(data['createdDate'])
      job.completed_at = DateTime.parse(data['systemModstamp'])
      job.external_id_field_name = data['externalIdFieldName']
      job.concurrency_mode = data['concurrencyMode']
      job.content_type = data['contentType']
      job.queued_batches = data['numberBatchesQueued'].to_i
      job.in_progress_batches = data['numberBatchesInProgress'].to_i
      job.completed_batches = data['numberBatchesCompleted'].to_i
      job.failed_batches = data['numberBatchesFailed'].to_i
      job.total_batches = data['numberBatchesTotal'].to_i
      job.retries = data['retries'].to_i
      job.processed_records = data['numberRecordsProcessed'].to_i
      job.failed_records = data['numberRecordsFailed'].to_i
      job.total_processing_time = data['totalProcessingTime'].to_i
      job.api_active_processing_time = data['apiActiveProcessingTime'].to_i
      job.apex_processing_time = data['apexProcessingTime'].to_i
      job.api_version = data['apiVersion'].to_i
      job
    end

    def aborted?
      state? 'Aborted'
    end

    def closed?
      state? 'Closed'
    end

    def open?
      state? 'Open'
    end

    def state?(value)
      !self.state.nil? && self.state.casecmp(value) == 0
    end
  end
end


lib/salesforce_bulk/batch.rb
module SalesforceBulk
  class Batch

    attr_accessor :apex_processing_time
    attr_accessor :api_active_processing_time
    attr_accessor :completed_at
    attr_accessor :created_at
    attr_accessor :failed_records
    attr_accessor :id
    attr_accessor :job_id
    attr_accessor :processed_records
    attr_accessor :state
    attr_accessor :state_message
    attr_accessor :total_processing_time

    def self.new_from_xml(data)
      batch = self.new
      batch.id = data['id']
      batch.job_id = data['jobId']
      batch.state = data['state']
      batch.state_message = data['stateMessage']
      batch.created_at = DateTime.parse(data['createdDate'])
      batch.completed_at = DateTime.parse(data['systemModstamp'])
      batch.processed_records = data['numberRecordsProcessed'].to_i
      batch.failed_records = data['numberRecordsFailed'].to_i
      batch.total_processing_time = data['totalProcessingTime'].to_i
      batch.api_active_processing_time = data['apiActiveProcessingTime'].to_i
      batch.apex_processing_time = data['apex_processing_time'].to_i
      batch
    end

    def in_progress?
      state? 'InProgress'
    end

    def queued?
      state? 'Queued'
    end

    def completed?
      state? 'Completed'
    end

    def failed?
      state? 'Failed'
    end

    def state?(value)
      !self.state.nil? && self.state.casecmp(value) == 0
    end
  end
end

lib/salesforce_bulk/salesforce_error.rb
module SalesforceBulk
  # An exception raised when any non successful request is made through the Salesforce Bulk API.
  class SalesforceError < StandardError
    # The Net::HTTPResponse instance from the API call.
    attr_accessor :response
    # The status code from the server response body.
    attr_accessor :error_code
    # The exception code from Salesforce's response
    attr_accessor :exception_code

    def initialize(response)
      self.response = response
      self.error_code = response.code

      if (300..308).cover?(error_code.to_i)
        message = "Moved to #{response['location']}"
      else
        data = XmlSimple.xml_in(response.body, 'ForceArray' => false)

        if data
          # seems responses for CSV requests use a different XML return format
          # (use invalid_error.xml for reference)
          if !data['exceptionMessage'].nil?
            message = data['exceptionMessage']
            self.exception_code = data['exceptionCode']
          else
            # SOAP error response
            message = data['Body']['Fault']['faultstring']
            self.exception_code = data['Body']['Fault']['faultcode']
          end
        end
      end

      super(message)
    end
  end
end


lib/salesforce_bulk/query_result_collection.rb
module SalesforceBulk
  class QueryResultCollection < Array

    attr_reader :client
    attr_reader :batch_id
    attr_reader :job_id
    attr_reader :result_id
    attr_reader :result_ids

    def initialize(client, job_id, batch_id, result_id=nil, result_ids=[])
      @client = client
      @job_id = job_id
      @batch_id = batch_id
      @result_id = result_id
      @result_ids = result_ids
      @current_index = result_ids.index(result_id) || 0
    end

    def next?
      !@result_ids.nil? && !@result_ids.empty? && @current_index < @result_ids.length - 1
    end

    def next
      if next?
        replace(@client.query_result(job_id, batch_id, result_ids[@current_index + 1]))
        @current_index += 1
        @result_id = @result_ids[@current_index]
      end

      self
    end

    def previous?
      !@result_ids.nil? && !@result_ids.empty? && @current_index > 0
    end

    def previous
      if previous?
        replace(@client.query_result(job_id, batch_id, result_ids[@current_index - 1]))
        @current_index -= 1
        @result_id = @result_ids[@current_index]
      end

      self
    end

  end
end

lib/salesforce_bulk/client.rb
module SalesforceBulk

  # Interface for operating the Salesforce Bulk REST API
  class Client
    # The host to use for authentication. Defaults to login.salesforce.com.
    attr_accessor :login_host

    # The instance host to use for API calls. Determined from login response.
    attr_accessor :instance_host

    # The Salesforce password
    attr_accessor :password

    # The Salesforce username
    attr_accessor :username

    # The API version the client is using. Defaults to 24.0.
    attr_accessor :version

    # The ID for authenticated session
    attr_accessor :session_id

    def initialize(options={})
      options = {:login_host => 'login.salesforce.com', :version => 24.0}.merge(options)

      assert_valid_keys(options, :username, :password, :login_host, :version, :instance_host, :session_id)

      self.username = options[:username]
      self.password = "#{options[:password]}"
      self.login_host = options[:login_host]
      self.version = options[:version]
      self.instance_host = options[:instance_host]
      self.session_id = options[:session_id]

      @api_path_prefix = "/services/async/#{version}/"
      @valid_operations = [:delete, :insert, :update, :upsert, :query]
      @valid_concurrency_modes = ['Parallel', 'Serial']
    end

    def authenticate
      # Clear session attributes just in case client already had a session
      @session_id = nil
      self.instance_host = nil

      xml = '<?xml version="1.0" encoding="utf-8"?>'
      xml += '<env:Envelope xmlns:xsd="http://www.w3.org/2001/XMLSchema"'
      xml += ' xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"'
      xml += ' xmlns:env="http://schemas.xmlsoap.org/soap/envelope/">'
      xml += "<env:Body>"
      xml += '<n1:login xmlns:n1="urn:partner.soap.sforce.com">'
      xml += "<n1:username>#{username}</n1:username>"
      xml += "<n1:password>#{password.encode(xml: :text)}</n1:password>"
      xml += "</n1:login>"
      xml += "</env:Body>"
      xml += "</env:Envelope>\n"

      response = http_post("/services/Soap/u/#{version}", xml, 'Content-Type' => 'text/xml', 'SOAPAction' => 'login')

      data = XmlSimple.xml_in(response.body, 'ForceArray' => false)
      result = data['Body']['loginResponse']['result']

      @session_id = result['sessionId']

      self.instance_host = "#{instance_id(result['serverUrl'])}.salesforce.com"
      self
    end

    def abort_job(jobId)
      xml = '<?xml version="1.0" encoding="utf-8"?>'
      xml += '<jobInfo xmlns="http://www.force.com/2009/06/asyncapi/dataload">'
      xml += "<state>Aborted</state>"
      xml += "</jobInfo>"

      response = http_post("job/#{jobId}", xml)
      data = XmlSimple.xml_in(response.body, 'ForceArray' => false)
      Job.new_from_xml(data)
    end

    def add_batch(jobId, data)
      body = data

      if data.is_a?(Array)
        raise ArgumentError, "Data set exceeds 10000 record limit by #{data.length - 10000}" if data.length > 10000

        keys = data.first.keys
        body = keys.to_csv

        data.each do |item|
          item_values = keys.map { |key| item[key] }
          body += item_values.to_csv
        end
      end

      # Despite the content for a query operation batch being plain text we
      # still have to specify CSV content type per API docs.
      response = http_post("job/#{jobId}/batch", body, "Content-Type" => "text/csv; charset=UTF-8")
      result = XmlSimple.xml_in(response.body, 'ForceArray' => false)
      Batch.new_from_xml(result)
    end

    def add_job(operation, sobject, options={})
      operation = operation.to_s.downcase.to_sym

      raise ArgumentError.new("Invalid operation: #{operation}") unless @valid_operations.include?(operation)

      assert_valid_keys(options, :external_id_field_name, :concurrency_mode)

      if options[:concurrency_mode]
        concurrency_mode = options[:concurrency_mode].capitalize
        raise ArgumentError.new("Invalid concurrency mode: #{concurrency_mode}") unless @valid_concurrency_modes.include?(concurrency_mode)
      end

      xml = '<?xml version="1.0" encoding="utf-8"?>'
      xml += '<jobInfo xmlns="http://www.force.com/2009/06/asyncapi/dataload">'
      xml += "<operation>#{operation}</operation>"
      xml += "<object>#{sobject}</object>"
      xml += "<externalIdFieldName>#{options[:external_id_field_name]}</externalIdFieldName>" if options[:external_id_field_name]
      xml += "<concurrencyMode>#{options[:concurrency_mode]}</concurrencyMode>" if options[:concurrency_mode]
      xml += "<contentType>CSV</contentType>"
      xml += "</jobInfo>"

      response = http_post("job", xml)
      data = XmlSimple.xml_in(response.body, 'ForceArray' => false)
      Job.new_from_xml(data)
    end

    def batch_info_list(jobId)
      response = http_get("job/#{jobId}/batch")
      result = XmlSimple.xml_in(response.body, 'ForceArray' => false)

      if result['batchInfo'].is_a?(Array)
        result['batchInfo'].collect do |info|
          Batch.new_from_xml(info)
        end
      else
        [Batch.new_from_xml(result['batchInfo'])]
      end
    end

    def batch_info(jobId, batchId)
      response = http_get("job/#{jobId}/batch/#{batchId}")
      result = XmlSimple.xml_in(response.body, 'ForceArray' => false)
      Batch.new_from_xml(result)
    end

    def batch_result(jobId, batchId)
      response = http_get("job/#{jobId}/batch/#{batchId}/result")

      if ['application/xml', 'text/xml'].include? response.content_type
        result = XmlSimple.xml_in(response.body)

        if !result['result'].nil? && !result['result'].empty?
          results = query_result(jobId, batchId, result['result'].first)

          collection = QueryResultCollection.new(self, jobId, batchId, result['result'].first, result['result'])
          collection.replace(results)
        end
      else
        result = BatchResultCollection.new(jobId, batchId)

        CSV.parse(response.body, :headers => true) do |row|
          result << BatchResult.new(row[0], to_boolean(row[1]), to_boolean(row[2]), row[3])
        end

        result
      end
    end

    def query_result(job_id, batch_id, result_id)
      headers = {"Content-Type" => "text/csv; charset=UTF-8"}
      response = http_get("job/#{job_id}/batch/#{batch_id}/result/#{result_id}", headers)

      lines = response.body.lines.to_a
      headers = CSV.parse_line(lines.shift).collect { |header| header.to_sym }

      result = []

      #CSV.parse(lines.join, :headers => headers, :converters => [:all, lambda{|s| to_boolean(s) if s.kind_of? String }]) do |row|
      CSV.parse(lines.join, :headers => headers) do |row|
        result << Hash[row.headers.zip(row.fields)]
      end

      result
    end

    def close_job(jobId)
      xml = '<?xml version="1.0" encoding="utf-8"?>'
      xml += '<jobInfo xmlns="http://www.force.com/2009/06/asyncapi/dataload">'
      xml += "<state>Closed</state>"
      xml += "</jobInfo>"

      response = http_post("job/#{jobId}", xml)
      data = XmlSimple.xml_in(response.body, 'ForceArray' => false)
      Job.new_from_xml(data)
    end

    def job_info(jobId)
      response = http_get("job/#{jobId}")
      data = XmlSimple.xml_in(response.body, 'ForceArray' => false)
      Job.new_from_xml(data)
    end

    def http_post(path, body, headers={})
      headers = {'Content-Type' => 'application/xml'}.merge(headers)

      if @session_id
        headers['X-SFDC-Session'] = @session_id
        host = instance_host
        path = "#{@api_path_prefix}#{path}"
      else
        host = self.login_host
      end

      response = https_request(host).post(path, body, headers)

      if response.is_a?(Net::HTTPSuccess)
        response
      else
        raise SalesforceError.new(response)
      end
    end

    def http_get(path, headers={})
      path = "#{@api_path_prefix}#{path}"

      headers = {'Content-Type' => 'application/xml'}.merge(headers)

      if @session_id
        headers['X-SFDC-Session'] = @session_id
      end

      response = https_request(self.instance_host).get(path, headers)

      if response.is_a?(Net::HTTPSuccess)
        response
      else
        raise SalesforceError.new(response)
      end
    end

    def https_request(host)
      req = Net::HTTP.new(host, 443)
      req.use_ssl = true
      req.verify_mode = OpenSSL::SSL::VERIFY_NONE
      req
    end

    def instance_id(url)
      url.match(/:\/\/([a-zA-Z0-9\-\.]{2,}).salesforce/)[1]
    end

    private

    def assert_valid_keys(options, *valid_keys)
      valid_keys.flatten!
      options.each_key do |k|
        unless valid_keys.include?(k)
          raise ArgumentError.new("Unknown key: #{k.inspect}. Valid keys are: #{valid_keys.map(&:inspect).join(', ')}")
        end
      end
    end

    def to_boolean(value)
      if !value.nil?
        if value.strip.casecmp("true") == 0
          return true
        elsif value.strip.casecmp("false") == 0
          return false
        end
      end
      value
    end
  end
end
